xds/internal/balancer/gracefulswitch/gracefulswitch.go

```diff
@@ -0,0 +1,399 @@
+	}
+	return gsb.balancerCurrent
+}
+
+// UpdateClientConnState forwards the update to the latest balancer created.
+func (gsb *Balancer) UpdateClientConnState(state balancer.ClientConnState) error {
+	balToUpdate := gsb.latestBalancer()
+	if balToUpdate == nil {
+		return errBalancerClosed
+	}
+
+	// Forwarding the update outside of holding gsb.mu prevents deadlock
+	// scenarios, as the update can induce an inline callback which requires
+	// gsb.mu. However, the current balancer can be closed at any time after
+	// giving up gsb.mu. The pending balancer cannot be closed, as we are
+	// guaranteed grpc will call the balancer API synchronously, and for the
+	// pending to also be closed would require another update from grpc. The
+	// current being closed is not a problem for this UpdateClientConnState()
+	// call, as in the scenario that the current balancer can be closed (current
+	// + pending balancers are both populated), this update will always be
+	// forwarded to the pending. Thus, there is a guarantee that this will not
+	// break the balancer API of the balancer by updating after closing.
+	return balToUpdate.UpdateClientConnState(state)
+}
+
+// ResolverError forwards the error to the latest balancer created.
+func (gsb *Balancer) ResolverError(err error) {
+	// The update will be forwarded to the pending balancer only if there is a
+	// pending LB present, as that is the most recent LB Config prepared by the
+	// resolver, and thus is a separate concern from the current.
+	balToUpdate := gsb.latestBalancer()
+	if balToUpdate == nil {
+		return
+	}
+	// Forwarding the update outside of holding gsb.mu prevents deadlock
+	// scenarios, as the update can induce an inline callback which requires
+	// gsb.mu. However, the current balancer can be closed at any time after
+	// giving up gsb.mu. The pending balancer cannot be closed, as we are
+	// guaranteed grpc will call the balancer API synchronously, and for the
+	// pending to also be closed would require another update from grpc. The
+	// current being closed is not a problem for this ResolverError() call, as
+	// in the scenario that the current balancer can be closed (current +
+	// pending balancers are both populated), this error will always be
+	// forwarded to the pending. Thus, there is a guarantee that this will not
+	// break the balancer API of the balancer by updating after closing.
+	balToUpdate.ResolverError(err)
+}
+
+// UpdateSubConnState forwards the update to the appropriate child.
+func (gsb *Balancer) UpdateSubConnState(sc balancer.SubConn, state balancer.SubConnState) {
+	// At any given time, the current balancer may be closed in the situation
+	// where there is a current balancer and a pending balancer. Thus, the
+	// swap() (close() + clearing of SubConns) operation is also guarded by this
+	// currentMu. Whether the current has been deleted or not will be checked
+	// from reading from the scToSubBalancer map, as that will be cleared in the
```

Comment:
`scToSugBalancerMap` was removed.  Please rewrite this comment; try to simplify.

Reply:
I went ahead and deleted this, as the explanation of this currentMu was added to both mu and currentMu.